xcdat/averager.py

```diff
@@ -0,0 +1,458 @@
+    return axis_weights
+
+
+def _combine_axis_weights(axis_weights, axis=['lat', 'lon']):
+    """Function generically re-scales axis weights for a given region.
+
+    Operations include:
+
+    - Create N-dimensional weighting array based on axis_weights and axis
+    dimensions specified
+
+    Parameters
+    ----------
+    axis_weights : dict
+        Dictionary of axis weights (using dims as keys)
+    axis : list, optional
+        List of axes that should be weighted (default ['lat', 'lon'])
+
+    Returns
+    -------
+    xr.DataArray
+        Array of weights
+
+    Examples
+    --------
+    Import and call module:
+    >>> from xcdat.dataset import open_dataset
+    >>> from xcdat.variable import open_variable
+    >>>
+    >>> filename = 'test.nc'
+    >>> ds = open_dataset(filename)
+    >>> tas = open_variable(ds, 'tas')
+    >>> axis_weights = _get_axis_weights(tas)
+    >>> weights = _combine_axis_weights(axis_weights, axis=['lat', 'lon'])
+    """
+    # ensure averaging axis is a list
+    if type(axis) == str:
+        axis = list(axis)
+    # initialize weight matrix
+    weights = axis_weights[axis[0]]
+    if len(axis) > 1:
+        for i in np.arange(1, len(axis)):
+            weights = weights * axis_weights[axis[i]]
+    return weights
+
+
+def _averager(xvariable, weights, axis):
+    """Perform a weighted average of a dataarray.
+
+    Operations include:
+
+    - Ensure masked data receives zero weight
+    - Perform weighted average over user-specified dimensions
+
+    Parameters
+    ----------
+    xvariable : xr.DataArray
+        xr.DataArray produced using ``open_variable``
+
+    weights : xr.DataArray
+        DataArray containing the desired weights for the variable array
+
+    axis : list
+        List of string values denoting the dims that should be averaged over
+        (e.g., axis=['lat', 'lon'])
+
+    Returns
+    -------
+    xr.DataArray
+        Dataarray that has been reduced via a weighted average.
+
+    Notes
+    -----
+    Needs to be evaluated for speed and compared to native xarray and CDAT
+    functionality.
+
+    """
+    # get masked data
+    masked = np.isnan(xvariable)
+    # ensure weights includes all of the same axes that are in the
+    # variable array
+    for i, dim in enumerate(xvariable.dims):
+        if dim not in weights.dims:
+            weights = weights.expand_dims(dim)
+    # tranpose to ensure axis ordering is the same
+    weights = weights.transpose(*xvariable.dims)
+    # expand array so that the variable and weights arrays have identical shape
+    # REVIEW: It may be useful to create a function (e.g., genutil.grower)
+    #         that does this operation.
+    ntile = []
+    for i, dim in enumerate(weights.dims):
+        if len(weights[dim]) != len(xvariable[dim]):
+            ntile.append(len(xvariable[dim]))
+        else:
+            ntile.append(1)
+    weights = np.tile(weights, ntile)
+    # zero out missing data
+    weights[masked] = 0.
```

Comment:
Can be replaced with http://xarray.pydata.org/en/stable/generated/xarray.DataArray.fillna.html